runtime/src/accounts_index.rs

```diff
@@ -204,30 +204,94 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
         })
     }
 
+    fn addref(item: &AtomicU64) {
+        item.fetch_add(1, Ordering::Relaxed);
+    }
+
+    pub fn upsert_new_key<'a>(
+        mut w_account_maps: AccountMapsWriteLock<'a, T>,
+        pubkey: &Pubkey,
+        new_value: AccountMapEntry<T>,
+        reclaims: &mut SlotList<T>,
+    ) {
+        match w_account_maps.entry(*pubkey) {
+            Entry::Occupied(mut occupied) => {
+                let current = occupied.get_mut();
+                Self::upsert_item(current, new_value, reclaims);
+            }
+            Entry::Vacant(vacant) => {
+                vacant.insert(new_value);
+            }
+        }
+    }
+
+    // returns None if upsert was successful
+    // otherwise, the pubkey has to be added to the maps with a write lock. call upsert_new
+    pub fn upsert_existing_key<'a>(
+        r_account_maps: AccountMapsReadLock<'a, T>,
+        pubkey: &Pubkey,
+        new_value: AccountMapEntry<T>,
+        reclaims: &mut SlotList<T>,
+    ) -> Option<AccountMapEntry<T>> {
+        if let Some(current) = r_account_maps.get(pubkey) {
+            Self::upsert_item(current, new_value, reclaims);
+            None
+        } else {
+            Some(new_value)
+        }
+    }
+
+    fn upsert_item(
```

Comment:
nit: `upsert_item` -> `lock_and_update_slot_list()` to distinguish from `update_slot_list` below.